nni/algorithms/hpo/gridsearch_tuner.py

```diff
@@ -2,209 +2,241 @@
+        self.epoch_bar = [0 for _ in self.space]
+        self.grid = [None for _ in self.space]
+        for i, spec in enumerate(self.space.values()):
+            if spec.categorical:
+                self.grid[i] = list(range(spec.size))
             else:
-                return self.expanded_search_space[self.count]
-        raise nni.NoMoreTrialError('no more parameters now.')
-
-    def receive_trial_result(self, parameter_id, parameters, value, **kwargs):
-        """
-        Receive a trial's final performance result reported through :func:`~nni.report_final_result` by the trial.
-        GridSearchTuner does not need trial's results.
-        """
-        pass
-
-    def import_data(self, data):
-        """
-        Import additional data for tuning
-
-        Parameters
-        ----------
-        list
-            A list of dictionarys, each of which has at least two keys, ``parameter`` and ``value``
-        """
-        _completed_num = 0
-        for trial_info in data:
-            logger.info("Importing data, current processing progress %s / %s", _completed_num, len(data))
-            _completed_num += 1
-            assert "parameter" in trial_info
-            _params = trial_info["parameter"]
-            assert "value" in trial_info
-            _value = trial_info['value']
-            if not _value:
-                logger.info("Useless trial data, value is %s, skip this trial data.", _value)
-                continue
-            _params_tuple = convert_dict2tuple(copy.deepcopy(_params))
-            self.supplement_data[_params_tuple] = True
-        logger.info("Successfully import data to grid search tuner.")
+                self.grid[i] = [0.5]
+                self.divisions[i] = []
+                if _less(0, 0.5, spec):
+                    self.divisions[i].append((0, 0.5))
+                if _less(0.5, 1, spec):
+                    self.divisions[i].append((0.5, 1))
+
+        size = _grid_size_info(self.grid)
+        _logger.info(f'Grid initialized, size: {size}')
+
+    def _current_parameters(self):
+        # convert self.vector to "formatted" parameters
+        params = {}
+        for i, spec in enumerate(self.space.values()):
+            x = self.grid[i][self.vector[i]]
+            if spec.categorical:
+                params[spec.key] = x
+            else:
+                params[spec.key] = _cdf_inverse(x, spec)
+        return params
+
+def _less(x, y, spec):
+    if spec.q is None:
+        return x < y
+    real_x = _deformat_single_parameter(_cdf_inverse(x, spec), spec)
+    real_y = _deformat_single_parameter(_cdf_inverse(y, spec), spec)
+    return real_x < real_y
+
+def _cdf_inverse(x, spec):
+    # inverse function of spec's cumulative distribution function
+    if spec.normal_distributed:
+        return spec.mu + spec.sigma * math.sqrt(2) * erfinv(2 * x - 1)
+    else:
+        return spec.low + (spec.high - spec.low) * x
+
+def _deformat_single_parameter(x, spec):
```

Comment:
Why is this needed provided that we have `deformat_parameters` already?

Reply:
Because it's not that easy to get a parameter inside nested space.
It either needs a function to do `while ...: obj = getattr(obj, ...)`, or this.
I think the complexity is similar, and chose this approach because it has better efficiency.